cmd/xl-zones.go

```diff
@@ -0,0 +1,1068 @@
+/*
+ * MinIO Cloud Storage, (C) 2019 MinIO, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"context"
+	"io"
+	"net/http"
+	"sort"
+	"strings"
+
+	xhttp "github.com/minio/minio/cmd/http"
+	"github.com/minio/minio/pkg/lifecycle"
+	"github.com/minio/minio/pkg/madmin"
+	"github.com/minio/minio/pkg/policy"
+	"github.com/minio/minio/pkg/sync/errgroup"
+)
+
+type xlZones struct {
+	zones []*xlSets
+}
+
+func (z *xlZones) SingleZone() bool {
+	return len(z.zones) == 1
+}
+
+// Initialize new zone of erasure codes.
+func newXLZones(endpointsList EndpointLists, formats []*formatXLV3) (ObjectLayer, error) {
+	z := &xlZones{}
+	for i, ep := range endpointsList {
+		sets, err := newXLSets(ep.Endpoints, formats[i], ep.SetCount, ep.DrivesPerSet)
+		if err != nil {
+			return nil, err
+		}
+		z.zones = append(z.zones, sets)
+	}
+	return z, nil
+}
+
+// AvailableZoneSlice attaches the methods of Uint64erface to []uint64, sorting in increasing order.
+type AvailableZoneSlice []struct {
+	Index     int
+	Available uint64
+}
+
+func (p AvailableZoneSlice) Len() int           { return len(p) }
+func (p AvailableZoneSlice) Less(i, j int) bool { return p[i].Available < p[j].Available }
+func (p AvailableZoneSlice) Swap(i, j int)      { p[i], p[j] = p[j], p[i] }
+
+func (z *xlZones) getLeastUsageZoneSlice(ctx context.Context) AvailableZoneSlice {
```

Comment:
You are sorting on least *available*, not least usage AFAICT.

The documentation states that `Available uint64 // Total disk space available.` on StorageInfo. So you are getting the reverse of what you expect.